src/index.js

```diff
@@ -93,7 +94,7 @@ export default function onClickOutsideHOC(WrappedComponent, config) {
       // If we are in an environment without a DOM such
       // as shallow rendering or snapshots then we exit
       // early to prevent any unhandled errors being thrown.
-      if (typeof document === 'undefined' || !document.createElement) {
+      if (typeof this.document === 'undefined' || !this.document.createElement) {
```

Comment:
given the assignment in the constructor, is `this.document` being `undefined` possible? It would probably be better to ensure that `this.document` is always a document, to the point where if L52 yields `null` or `undefined`, we assign a duck-typed document-like object that has exactly as much as the HOC needs to at least think it's working, and then warn about the fact that we had to do that with a `console.warn` or `console.error`.

This removes a whole bunch of checks the result of which should already be known at the end of running the constructor (since we can't move components from one document to another, even if we could conceivably serialize/deserialize across documents)

Reply:
> given the assignment in the constructor, is this.document being undefined possible?

Yeah, it can be undefined (or rather a `null` with my proposed change in the other comment) in case of SSR. We should leave those ifs here.

We could OTOH warn in development mode about returning "broken" document from `getDocument` - we'd have to detect being in a browser though, to skip warning in case of SSR too.

It would look something like this:
```js
const isBrowser = () => typeof document !== "undefined" && document.createElement

// ...

this.document = getDocument(this.getInstance())
if (process.env.NODE_ENV !== 'production' && getDocument !== defaultGetDocument && typeof this.document.addEventListener !== 'function') {
  // or maybe even we should fail fast here with throw new Error?
  console.warn('WRITE A WARNING HERE')
}
```

Reply:
Also this if here:
```js
if (typeof this.document === 'undefined' || !this.document.createElement) {
```
should not be about checking if w have a valid `this.document` but rather about checking if we are in the browser env

Reply:
@Andarist  `isBrowser()` makes sense. So in the constructor, we can just assign document as null if its not a browser env

````
this.document = isBrowser()
        ? config && config.getDocument
          ? config.getDocument(this.getInstance())
          : document
        : null;
````

And all `this.document` checks can be replaced with `isBrowser()` check? Wouldn't that suffice? Test are passing

Reply:
yeah, something like that should work, please use default `getDocument` though, will make this clearer (`const getDocument = () => document`)